File: Tec_Prog/1-Factory-Methood/Exemplos/FactoryMethood.js
//Classe base veiculo
class Veiculo {
    constructor(modelo){
        this.modelo = modelo
    }
 
    mostrarDetalhes(){
        console.log(`Modelo: ${this.modelo}`)
    }
}
 
//Subclasse Veiculo
class Carro extends Veiculo{
    constructor(modelo){
        super(modelo);
    };
};
 
class Moto extends Veiculo{
    constructor(modelo){
        super(modelo);
    }
}
 
class Barco extends Veiculo{
    constructor(modelo){
        super(modelo)
    }
}
 
//Fabrica abstrata de veiculo
class FabricadeVeiculos{
    criarVeiculo(modelo){
        throw new Error('O método criar veiculo deve ser implementado')
    }
}
//fabrica concreta de carros
class FabricaDeCarros extends FabricadeVeiculos{
    criarVeiculo(modelo){
        return new Carro(modelo)
    }
}
 
//Fabrica concreta de motos
class FabricaDeMotos extends FabricadeVeiculos{
    criarVeiculo(modelo){
        return new Moto(modelo)
    }
}
 
//Fabrica concreta de barcos
class FabricaDeBarcos extends FabricadeVeiculos{
    criarVeiculo(modelo){
        return new Barco(modelo)
    }
}
 
//Uso de fabrica cliente
const fabricaCarros = new FabricaDeCarros();
const fabricaMotos = new FabricaDeMotos();
const fabricaBarcos = new FabricaDeBarcos();
 
const veiculo1 = fabricaCarros.criarVeiculo('Sedan')
const veiculo2 = fabricaCarros.criarVeiculo('Hatch')
const veiculo3 = fabricaMotos.criarVeiculo('Esportiva')
const veiculo4 = fabricaMotos.criarVeiculo('Custon')
const veiculo5 = fabricaBarcos.criarVeiculo('Titanic')
 
veiculo1.mostrarDetalhes()
veiculo2.mostrarDetalhes()
veiculo5.mostrarDetalhes()
 

File: Tec_Prog/1-Factory-Methood/Exemplos/FactoryMethood.java
// Classe Base  Veiculo
class Veiculo{
    private String modelo;
 
    Veiculo(String modelo){
        this.modelo = modelo;
    }
 
    public void mostrarDetalhes(){
        System.out.println("Modelo: " + modelo);
    }
}
 
class Carro extends Veiculo{
    public Carro(String modelo){
        super(modelo);
    }
}
 
class Moto extends Veiculo{
    public Moto(String modelo){
        super(modelo);
    }
}
 
class Barco extends Veiculo{
    public Barco(String modelo){
        super(modelo);
    }
}
 
abstract class FabricaDeVeiculos{
    public abstract criarVeiculo(String modelo);
}
 
class FabricaDeCarros extends FabricaDeVeiculos{
    public Veiculo criarVeiculo(String modelo){
        return new Carro(modelo);
    }
}
 
class FabricaDeMotos extends FabricaDeVeiculos{
    public Veiculo criarVeiculo(String modelo){
        return new Moto(modelo);
    }
}
 
class FabricaDeBarcos extends FabricaDeVeiculos{
    public Veiculo criarVeiculo(String modelo){
        return new Moto(modelo);
    }
}
 
public class FactoryMethood03{
    public static void main(String[] args){
        FabricaDeVeiculos fabricaCarros = new FabricaDeCarros();
        FabricaDeMotos fabricaMotos = new FabricaDeMotos();
 
        fabricaCarros.criarVeiculo('Fox');
        fabricaMotos.criarVeiculo('Honda');
 
        fabricaMotos.mostrarDetalhes();
        fabricaMotos.mostrarDetalhes();
    }
}